**Código da aula 3:**

In [5]:
import pandas as pd

# Caminho do arquivo
filepath = '/content/student-mat.csv'

# Ler o arquivo CSV
df = pd.read_csv(filepath, delimiter=';')

# Visualizar as primeiras linhas para entender a estrutura dos dados
df.head()


school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        6   5   6   6  
1      5        3      3     1     1      3        4   5   5   6  
2      4        3      2     2     3      3       10   7   8  10  
3      3        2      2     1     1      5        2  15  14  15  
4      4        3      2     1     2      5        4   6  10  10  

[5 rows x 33 columns]

In [6]:
# Verificar as informações básicas do dataset
info = df.info()

# Análise descritiva dos dados
desc = df.describe()

info, desc


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      395 non-null    object
 1   sex         395 non-null    object
 2   age         395 non-null    int64 
 3   address     395 non-null    object
 4   famsize     395 non-null    object
 5   Pstatus     395 non-null    object
 6   Medu        395 non-null    int64 
 7   Fedu        395 non-null    int64 
 8   Mjob        395 non-null    object
 9   Fjob        395 non-null    object
 10  reason      395 non-null    object
 11  guardian    395 non-null    object
 12  traveltime  395 non-null    int64 
 13  studytime   395 non-null    int64 
 14  failures    395 non-null    int64 
 15  schoolsup   395 non-null    object
 16  famsup      395 non-null    object
 17  paid        395 non-null    object
 18  activities  395 non-null    object
 19  nursery     395 non-null    object
 20  higher    

(None,
               age        Medu        Fedu  traveltime   studytime    failures  \
 count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
 mean    16.696203    2.749367    2.521519    1.448101    2.035443    0.334177   
 std      1.276043    1.094735    1.088201    0.697505    0.839240    0.743651   
 min     15.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
 25%     16.000000    2.000000    2.000000    1.000000    1.000000    0.000000   
 50%     17.000000    3.000000    2.000000    1.000000    2.000000    0.000000   
 75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
 max     22.000000    4.000000    4.000000    4.000000    4.000000    3.000000   
 
            famrel    freetime       goout        Dalc        Walc      health  \
 count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
 mean     3.944304    3.235443    3.108861    1.481013    2.291139    3.554430   
 std   

In [11]:
pd.set_option('display.precision', 2)
df.describe()

age    Medu    Fedu  traveltime  studytime  failures  famrel  \
count  395.00  395.00  395.00      395.00     395.00    395.00  395.00   
mean    16.70    2.75    2.52        1.45       2.04      0.33    3.94   
std      1.28    1.09    1.09        0.70       0.84      0.74    0.90   
min     15.00    0.00    0.00        1.00       1.00      0.00    1.00   
25%     16.00    2.00    2.00        1.00       1.00      0.00    4.00   
50%     17.00    3.00    2.00        1.00       2.00      0.00    4.00   
75%     18.00    4.00    3.00        2.00       2.00      0.00    5.00   
max     22.00    4.00    4.00        4.00       4.00      3.00    5.00   

       freetime   goout    Dalc    Walc  health  absences      G1      G2  \
count    395.00  395.00  395.00  395.00  395.00    395.00  395.00  395.00   
mean       3.24    3.11    1.48    2.29    3.55      5.71   10.91   10.71   
std        1.00    1.11    0.89    1.29    1.39      8.00    3.32    3.76   
min        1.00    1.00    1.00    1.00    1.00      0.00    3.00    0.00   
25%        3.00    2.00    1.00    1.00    3.00      0.00    8.00    9.00   
50%        3.00    3.00    1.00    2.00    4.00      4.00   11.00   11.00   
75%        4.00    4.00    2.00    3.00    5.00      8.00   13.00   13.00   
max        5.00    5.00    5.00    5.00    5.00     75.00   19.00   19.00   

           G3  
count  395.00  
mean    10.42  
std      4.58  
min      0.00  
25%      8.00  
50%     11.00  
75%     14.00  
max     20.00

In [12]:
# Remover linhas com valores ausentes
data_clean = data.dropna()

# Verificar se ainda há valores ausentes e exibir as primeiras linhas após a limpeza
data_clean.isnull().sum(), data_clean.head()


(school        0
 sex           0
 age           0
 address       0
 famsize       0
 Pstatus       0
 Medu          0
 Fedu          0
 Mjob          0
 Fjob          0
 reason        0
 guardian      0
 traveltime    0
 studytime     0
 failures      0
 schoolsup     0
 famsup        0
 paid          0
 activities    0
 nursery       0
 higher        0
 internet      0
 romantic      0
 famrel        0
 freetime      0
 goout         0
 Dalc          0
 Walc          0
 health        0
 absences      0
 G1            0
 G2            0
 G3            0
 dtype: int64,
   school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
 0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
 1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
 2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
 3     GP   F   15       U     GT3       T     4     2   health  services  ...   
 4     GP   F 

In [7]:
# Remover instâncias duplicadas
df = df.drop_duplicates()

# Selecionar características para o modelo
features = df.columns.drop(['G2', 'G3'])  # Exclui G2 e G3 das características
X_possiveis = df[features]
y = df['G3']  # Variável de saída

# Verificar a nova estrutura dos dados após remoção de duplicatas e seleção de características
X_possiveis.head(), y.head()


(  school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
 0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
 1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
 2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
 3     GP   F   15       U     GT3       T     4     2   health  services  ...   
 4     GP   F   16       U     GT3       T     3     3    other     other  ...   
 
   internet romantic  famrel  freetime  goout Dalc Walc health absences  G1  
 0       no       no       4         3      4    1    1      3        6   5  
 1      yes       no       5         3      3    1    1      3        4   5  
 2      yes       no       4         3      2    2    3      3       10   7  
 3      yes      yes       3         2      2    1    1      5        2  15  
 4       no       no       4         3      2    1    2      5        4   6  
 
 [5 rows x 31 columns],
 0     6
 1 

In [8]:
# Identificar colunas categóricas
colunas_categoricas = X_possiveis.select_dtypes(include=['object']).columns

# Aplicar one-hot encoding às variáveis categóricas
X_encoded = pd.get_dummies(X_possiveis, columns=colunas_categoricas)

# Visualizar as primeiras linhas do dataframe após o one-hot encoding
X_encoded.head()


age  Medu  Fedu  traveltime  studytime  failures  famrel  freetime  goout  \
0   18     4     4           2          2         0       4         3      4   
1   17     1     1           1          2         0       5         3      3   
2   15     1     1           1          2         3       4         3      2   
3   15     4     2           1          3         0       3         2      2   
4   16     3     3           1          2         0       4         3      2   

   Dalc  ...  activities_no  activities_yes  nursery_no  nursery_yes  \
0     1  ...              1               0           0            1   
1     1  ...              1               0           1            0   
2     2  ...              1               0           0            1   
3     1  ...              0               1           0            1   
4     1  ...              1               0           0            1   

   higher_no  higher_yes  internet_no  internet_yes  romantic_no  romantic_yes  
0          0           1            1             0            1             0  
1          0           1            0             1            1             0  
2          0           1            0             1            1             0  
3          0           1            0             1            0             1  
4          0           1            1             0            1             0  

[5 rows x 57 columns]

In [9]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler

# Dividir os dados em conjuntos de treinamento e teste
X_train, X_test, y_train, y_test = train_test_split(X_encoded, y, test_size=0.2, random_state=42)

# Normalizar os dados
scaler = MinMaxScaler()
scaler.fit(X_train)  # Ajusta o scaler apenas nos dados de treinamento

X_train_scaled = scaler.transform(X_train)  # Aplica a transformação nos dados de treinamento
X_test_scaled = scaler.transform(X_test)  # Aplica a mesma transformação nos dados de teste

# Verificar as primeiras linhas dos dados de treinamento normalizados para garantir que a normalização foi aplicada
X_train_scaled[:5]


array([[0.14285714, 0.75      , 0.75      , 0.        , 0.33333333,
        0.        , 0.75      , 0.25      , 0.5       , 0.        ,
        0.25      , 0.5       , 0.02666667, 0.5       , 1.        ,
        0.        , 0.        , 1.        , 0.        , 1.        ,
        1.        , 0.        , 0.        , 1.        , 0.        ,
        0.        , 0.        , 1.        , 0.        , 0.        ,
        0.        , 1.        , 0.        , 0.        , 0.        ,
        1.        , 0.        , 0.        , 0.        , 1.        ,
        0.        , 1.        , 0.        , 1.        , 0.        ,
        0.        , 1.        , 0.        , 1.        , 0.        ,
        1.        , 0.        , 1.        , 0.        , 1.        ,
        0.        , 1.        ],
       [0.14285714, 0.5       , 0.75      , 0.33333333, 0.        ,
        0.        , 1.        , 0.5       , 0.5       , 0.        ,
        0.        , 0.5       , 0.        , 0.57142857, 1.        ,
        0.     

In [10]:
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error, r2_score

# Inicializar os modelos
model_linear = LinearRegression()
model_tree = DecisionTreeRegressor(random_state=42)

# Treinar os modelos
model_linear.fit(X_train_scaled, y_train)
model_tree.fit(X_train_scaled, y_train)

# Fazer previsões nos conjuntos de teste
y_pred_linear = model_linear.predict(X_test_scaled)
y_pred_tree = model_tree.predict(X_test_scaled)

# Avaliar os modelos
mse_linear = mean_squared_error(y_test, y_pred_linear)
mse_tree = mean_squared_error(y_test, y_pred_tree)

r2_linear = r2_score(y_test, y_pred_linear)
r2_tree = r2_score(y_test, y_pred_tree)

mse_linear, mse_tree, r2_linear, r2_tree


(8.408623417721518, 10.49367088607595, 0.5899242119369863, 0.48823961491576284)

Regressão Linear:

Mean Squared Error (MSE): 8.38

Coeficiente de Determinação (R²): 0.59

Árvore de Decisão para Regressão:

Mean Squared Error (MSE): 10.49

Coeficiente de Determinação (R²): 0.49


A Regressão Linear foi um pouco melhor que a Árvore de Decisão ao prever as notas finais, pois teve erros menores.
Para este caso, a Regressão Linear pode ser mais eficaz.